examples/ChefPolicyfileWorkflow.md

```diff
@@ -0,0 +1,893 @@
+means to inject secrets rather than relying upon setting and storing them as
+attributes.
+
+### Method 2: Attribute Cookbook
+In this method, you create a cookbook that has a 1:1 relationship with the 
+Policyfile you've created. The logic is as follows:
+
+#### Process
+* Create a Policyfile Cookbook and use the generated cookbook to store 
+attributes. This is the cookbook that's defined like so in the `Policyfile.rb`.
+    ```
+    cookbook 'your-policy-name', path: '.'
+    ```
+* Define your attributes in the `attributes/default.rb` just like you would any
+ordinary cookbook.
+* Use `case` statements to define attributes based on Policy Groups, like so
+  note that `local` is the default Policy Group used by Test Kitchen:
+    ```
+    case policy_group
+        when 'local'
+            default['mongodb']['config']['mongod']['net']['port'] = 27017
+        when 'dev'
+            default['mongodb']['config']['mongod']['net']['port'] = 27018
+        when 'qa'
+            default['mongodb']['config']['mongod']['net']['port'] = 27019
+        when 'prod'
+            default['mongodb']['config']['mongod']['net']['port'] = 27020
+    end
+    ```
+    In the cookbook, I would reference the above attribute as:
+    ```
+    node['mongodb']['config']['mongod']['net']['port']
+    ```
+* Define your recipe logic in the Policyfile Cookbook and use it as the only
+cookbook in your runlist. This ensures it is used as a "wrapper cookbook" and
+that all attributes you have defined will take precedence
+
+#### Pros
+* You can keep your attributes defined in a separate file from your Policyfile,
+which can lead to cleaner code.
+* You can have external attribute sources that you pull in to multiple Policies
+at build time (removes the need for repeating everything).
+
+#### Cons/Gotchas
+* A few more steps involved than using Method 2.
+* If your policy has an `include_policy`, and the included policy has a default
+attribute defined that you want to override, you must use `override` when 
+defining your attribute. For example, my base policy might have an attribute
+cookbook that defins this:
+    ```
+    case policy_group
+        when 'local'
+            default['effortless']['audit']['user_toml']['interval'] = 1800
+    end
+    ```
+    If I want to set the value to `900` in my downstream policy, I need to add
+    this to my attributes:
+    ```
+    case policy_group
+        when 'local'
+            override['effortless']['audit']['user_toml']['interval'] = 900
+    end
+    ```
+    This may not be obvious at first because if you don't set `override`, 
+    there's a possibility that what you set as default will take precedence.
+    However, there are scenarios where that's not the case, so it's best to be
+    100% sure and set them with `override`. For all other attributes that are
+    not defined in an "included policy", you can set them just like any other
+    by using `default`.
```

Comment:
"its best to always set values with override" i'd classify as a clear design smell.

i'm not sure what the use case is where that's necessary offhand, but it may also be a bit of an indication that the `include_policy` mechanism generally is engineered as a bit of a hack.

that's the kind of thing that would be good to surface to a design issue in chef/chef because if using include_policy pushes people to having to override attributes, either there needs to be design principles that work around it (rather than embracing it) or else we need to think about engineering something.

Reply:
Thanks to some help from a customer, I've identified the times when we need to use `override` attributes, it's when an attribute is defined as an array in the upstream included policy, and then modified in the downstream. If `default` is used, then the array is appended to, when `override` is used, it is replaced entirely. This may or may not be intended behavior, it's hard to tell due to the lack of documentation on the matter.

The benefits of using `include_policy` seem to outweigh the risks in this case. I'm going to go ahead and create an issue in chef/chef to at least get a better understanding of "normal" behavior in this case, and whether a fix is warranted.

🍻